From ede9abea7abc1fbee616dd9894ce172c11e25d50 Mon Sep 17 00:00:00 2001
From: "christian.mardmoeller" <christian.mardmoeller@renesas.com>
Date: Thu, 14 Apr 2022 11:57:37 +0000
Subject: [PATCH 052/151] Correct DT for 2G5 and RH. Improved SERDES init for
 flexible interface up sequence. The function is checked on all 1G interfaces
 with new names sw0p0, sw0p1, sw0p2. Data transmission on 2G5 not yet
 confirmed but link and serdes comes up.

---
 .../renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso    |   2 +
 .../renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso    |   1 +
 arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts  |   3 +
 .../ethernet/renesas/rswitch2/rswitch2_eth.c  | 474 +++++++++++++-----
 .../ethernet/renesas/rswitch2/rswitch2_eth.h  |   1 +
 .../renesas/rswitch2/rswitch2_platf.c         |  26 +-
 6 files changed, 367 insertions(+), 140 deletions(-)

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso
index 3cd077be73f5..7da8fee570f9 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso
@@ -41,6 +41,8 @@
 
 &tsn0 {
 	phy-handle = <&tsn0_phy_2g5>;
+	phy-mode = "usxgmii";
+
 };
 
 &tsn0_phy_2g5 {
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso
index aba41c950217..e3b1cfc88d5e 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso
@@ -32,6 +32,7 @@
 
 &tsn1 {
 	phy-handle = <&tsn1_phy_2g5>;
+	phy-mode = "usxgmii";
 };
 
 &tsn1_phy_2g5 {
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
index aabf78f38167..f9291a122bf2 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
@@ -274,6 +274,7 @@
 			/* phy-handle set via plugin */
 			tsn0_phy_2g5: ethernet-phy@0 {
 				reg = <0>;
+				compatible = "ethernet-phy-ieee802.3-c45";
 				status = "disabled";
 			};
 			tsn0_phy_1g: ethernet-phy@3 {
@@ -287,6 +288,7 @@
 			/* phy-handle set via plugin */
 			tsn1_phy_2g5: ethernet-phy@0 {
 				reg = <0>;
+				compatible = "ethernet-phy-ieee802.3-c45";
 				status = "disabled";
 			};
 			tsn1_phy_1g: ethernet-phy@3 {
@@ -300,6 +302,7 @@
 			/* phy-handle set via plugin */
 			tsn2_phy_1g_rh: ethernet-phy@1 {
 				reg = <1>;
+				compatible = "ethernet-phy-ieee802.3-c45";
 				status = "disabled";
 			};
 			tsn2_phy_1g: ethernet-phy@3 {
diff --git a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
index 57bcc8e0d770..0bd0bda833c4 100644
--- a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
+++ b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
@@ -4,6 +4,8 @@
  * Copyright (C) 2019-2021 Renesas Electronics Corporation
  *
  */
+#define DEBUG_SERDES
+
 #include <linux/clk.h>
 #include <linux/delay.h>
 #include <linux/dma-mapping.h>
@@ -61,9 +63,11 @@ static int rswitch2_emac_set_state(struct net_device *ndev, enum emac_op state)
 						reg_val & FIELD_PREP(EAMS_OPS, state),
 						RSWITCH2_REG_POLL_DELAY, RSWITCH2_REG_POLL_TIMEOUT);
 	if (ret != 0) {
-		netdev_err(ndev, "Setting EMAC state timed out\n");
+		netdev_err(ndev, "Setting EMAC state to %d timed out\n", reg_val);
 		return ret;
 	}
+	printk("New ETHA mode is EAMC=%x  EMAS=%x  expected=%x\n",
+		ioread32(etha_base_addr + RSW2_ETHA_EAMC), ioread32(etha_base_addr + RSW2_ETHA_EAMS), state);
 
 	return 0;
 }
@@ -205,14 +209,14 @@ static int rswitch2_phy_init(struct net_device *ndev, unsigned int port_num)
 			const char *prop_phy_mode_str;
 
 			ret = of_property_read_u32(port, "reg", &port_reg);
-			if(ret < 0) {
+			if (ret < 0) {
 				dev_err(rsw2->dev, "Failed to get register property for port %d: %d\n",
 						cur_port_num, ret);
 				break;
 			}
-			else {
-				dev_info(rsw2->dev, "Got reg=%u for port %d\n", port_reg, port_num);
-			}
+			//else {
+			//	dev_info(rsw2->dev, "Got reg=%u for port %d\n", port_reg, port_num);
+			//}
 
 			ret = of_property_read_string(port, "phy-mode", &prop_phy_mode_str);
 			if(ret < 0) {
@@ -221,8 +225,8 @@ static int rswitch2_phy_init(struct net_device *ndev, unsigned int port_num)
 				break;
 			}
 
-			if(port_reg == port_num) {
-				dev_err(rsw2->dev, "Got matching port!\n");
+			if (port_reg == port_num) {
+				//dev_err(rsw2->dev, "Got matching port!\n");
 				break;
 			}
 		}
@@ -233,7 +237,7 @@ static int rswitch2_phy_init(struct net_device *ndev, unsigned int port_num)
 			return -ENODEV;
 		}
 		else {
-			dev_info(rsw2->dev, "Got phy_iface: %d\n", phy_port->phy_iface);
+			dev_info(rsw2->dev, "Got phy_iface mode: %s (%d)\n", phy_modes(phy_port->phy_iface), phy_port->phy_iface);
 		}
 
 		/* Try connecting to PHY */
@@ -281,9 +285,9 @@ static int rswitch2_phy_init(struct net_device *ndev, unsigned int port_num)
 			dev_info(rsw2->dev, "Connected: phydev is at 0x%px\n", phydev);
 		}
 		of_node_put(pn);
-		dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
+		//dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 		of_node_put(ports);
-		dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
+		//dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 
 	} else {
 		/* In non device implementation, we get the port data passed this way */
@@ -294,8 +298,10 @@ static int rswitch2_phy_init(struct net_device *ndev, unsigned int port_num)
 		netdev_dbg(ndev, "No device tree information. Getting PHY from ID\n");
 
 		phy_port->phy_iface = port_data->phy_iface;
+		dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 		phydev = phy_connect(ndev, port_data->phy_id, &rswitch2_phy_state_change,
 							phy_port->phy_iface);
+		dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 	}
 
 	if (IS_ERR(phydev)) {
@@ -314,65 +320,136 @@ static int rswitch2_phy_init(struct net_device *ndev, unsigned int port_num)
 /* SerDes functionality */
 void rswitch2_serdes_write32(void __iomem *addr, u32 offs,  u32 bank, u32 data)
 {
+#ifdef DEBUG_SERDES
+	printk("%s(%d, %03x : %04x = %04x)\n", __FUNCTION__, (int)((u64)addr&0xc00)/0x400, bank, offs, data);
+#endif
 	iowrite32(bank, addr + RSWITCH_SERDES_BANK_SELECT);
 	iowrite32(data, addr + offs);
 }
 
 u32 rswitch2_serdes_read32(void __iomem *addr, u32 offs,  u32 bank)
 {
+	u32 ret;
 	iowrite32(bank, addr + RSWITCH_SERDES_BANK_SELECT);
-	{
-		u32 ret = ioread32(addr + offs);
-		return ret;
-	}
+	ret = ioread32(addr + offs);
+#ifdef DEBUG_SERDES
+	printk("%s(%d, %03x : %04x) is %04x\n", __FUNCTION__, (int)((u64)addr&0xc00)/0x400, bank, offs, ret);
+#endif
+	return ret;
 }
 
 static int rswitch2_serdes_reg_wait(void __iomem *addr, u32 offs, u32 bank, u32 mask, u32 expected)
 {
+	u32 ret;
 	int i;
 
 	iowrite32(bank, addr + RSWITCH_SERDES_BANK_SELECT);
-	udelay(100);
+	//udelay(100);
 
-	for (i = 0; i < RSWITCH2_REG_POLL_TIMEOUT; i++) {
-		if ((ioread32(addr + offs) & mask) == expected)
+	for (i = 0; i < 1000 /*RSWITCH2_REG_POLL_TIMEOUT*/; i++) {
+		ret = ioread32(addr + offs);
+		if ((ret & mask) == expected) {
+#ifdef DEBUG_SERDES
+			printk("%s(%d, %03x : %04x) is %04x after %d iterations\n", __FUNCTION__, (int)((u64)addr&0xc00)/0x400, bank, offs, ret, i);
+#endif
 			return 0;
-		udelay(1);
+		}
+		mdelay(1);
 	}
-
+#ifdef DEBUG_SERDES
+	printk("%s(%d, %03x : %04x) is %04x but %04x==%04x expected\n", __FUNCTION__, (int)((u64)addr&0xc00)/0x400, bank, offs, ret, (ret&mask), expected);
+#endif
 	return -ETIMEDOUT;
 }
 
 static int rswitch2_serdes_init_sram(struct rswitch2_drv *rsw2)
 {
 	int ret;
+	int lane;
+	void __iomem *addr;
 
-	ret = rswitch2_serdes_reg_wait(rsw2->serdes_base_addr, VR_XS_PMA_MP_12G_16G_25G_SRAM,
-				      BANK_180, BIT(0), 0x01);
-	if (ret)
-		return ret;
+	//printk("%s()\n", __FUNCTION__);
+	for (lane = 0; lane < rsw2->num_of_tsn_ports; lane++) {
+		addr = rsw2->serdes_base_addr + (RSW2_SERDES_CHANNEL_OFFSET * lane);
+		ret = rswitch2_serdes_reg_wait(addr,
+				VR_XS_PMA_MP_12G_16G_25G_SRAM, BANK_180, BIT(0), 0x01);
+		if (ret) {
+			dev_err(rsw2->dev, "SERDES SRAM init on lane %d failed (step 1)\n", lane);
+			return ret;
+		}
+break;
+	}
 
+	//Step 2 only on lane 0
 	rswitch2_serdes_write32(rsw2->serdes_base_addr, VR_XS_PMA_MP_12G_16G_25G_SRAM,
 			       BANK_180, 0x3);
 
-	//CMA: Why not BANK_380 as in manual?
-	ret = rswitch2_serdes_reg_wait(rsw2->serdes_base_addr, SR_XS_PCS_CTRL1, BANK_300,
-				      BIT(15), 0);
-	if (ret)
-		return ret;
-
+	for (lane = 0; lane < rsw2->num_of_tsn_ports; lane++) {
+		addr = rsw2->serdes_base_addr + (RSW2_SERDES_CHANNEL_OFFSET * lane);
+		ret = rswitch2_serdes_reg_wait(addr,
+				SR_XS_PCS_CTRL1, BANK_300, BIT(15), 0);
+		if (ret) {
+			dev_err(rsw2->dev, "SERDES on lane %d does not finish reset (step 3)\n", lane);
+			return ret;
+		}
+break;
+	}
 	return 0;
 }
 
 static int rswitch2_serdes_common_setting(struct rswitch2_eth_port *eth_port)
 {
-	struct rswitch2_drv *rsw2;
+	int ret;
+	int lane;
 	void __iomem *addr;
+	struct rswitch2_drv *rsw2;
 
 	rsw2 = eth_port->rsw2;
-	addr = rsw2->serdes_base_addr;
+	//printk("%s(port=%d)\n", __FUNCTION__, eth_port->port_num);
+
+	/* Disable FUSE_OVERRIDE_EN */
+	for (lane = 0; lane < rsw2->num_of_tsn_ports; lane++) {
+		addr = rsw2->serdes_base_addr + RSWITCH_SERDES_FUSE_OVERRIDE(lane);
+		if (ioread32(addr))
+			iowrite32(0, addr);
+	}
+
+	/* Initialize SRAM */
+	ret = rswitch2_serdes_init_sram(rsw2);
+	if (ret)
+		return ret;
+
+	//printk("Step 2\n");
+	for (lane = 0; lane < rsw2->num_of_tsn_ports; lane++) {
+		addr = rsw2->serdes_base_addr + (RSW2_SERDES_CHANNEL_OFFSET * lane);
+		rswitch2_serdes_write32(addr, VR_XS_PCS_SFTY_MR_CTRL, BANK_380, 0x443);
+break;
+	}
 
 	//to allow mix operation SGMII and USXGMII PLLs configured both
+	addr = rsw2->serdes_base_addr;
+#if 0
+	//switch (mode) {
+	//case SGMII:
+		//S4.1~S4.5
+		rswitch_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_REF_CLK_CTRL, BANK_180, 0x97);
+		rswitch_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_MPLLB_CTRL0, BANK_180, 0x60);
+		rswitch_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_MPLLB_CTRL2, BANK_180, 0x2200);
+		rswitch_serdes_write32(addr, VR_XS_PMA_MP_12G_MPLLB_CTRL1, BANK_180, 0);
+		rswitch_serdes_write32(addr, VR_XS_PMA_MP_12G_MPLLB_CTRL3, BANK_180, 0x3d);
+		break;
+
+#elif 0
+	//case USXGMII:
+		//U4.1 ~ U4.5
+		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_REF_CLK_CTRL, BANK_180, 0x57);
+		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_10G_MPLLA_CTRL2, BANK_180, 0xc200);
+		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_MPLLA_CTRL0, BANK_180, 0x42);
+		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_MPLLA_CTRL1, BANK_180, 0);
+		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_MPLLA_CTRL3, BANK_180, 0x2f);
+		//break;
+
+#else
 	//C4.1 == U4.1 | S4.1
 	rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_REF_CLK_CTRL, BANK_180, 0x57 | 0x97);
 
@@ -393,8 +470,31 @@ static int rswitch2_serdes_common_setting(struct rswitch2_eth_port *eth_port)
 	rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_MPLLB_CTRL1, BANK_180, 0);
 	//C4.9 == S4.2
 	rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_MPLLB_CTRL3, BANK_180, 0x3d);
+#endif
+	//default:
+	//	return -EOPNOTSUPP;
+	//}
+
+	/* Assert softreset for SERDES PHY (step 5) */
+	for (lane = 0; lane < rsw2->num_of_tsn_ports; lane++) {
+		addr = rsw2->serdes_base_addr + (RSW2_SERDES_CHANNEL_OFFSET * lane);
+		rswitch2_serdes_write32(addr, 0x03d0, BANK_380, 1);
+break;
+	}
+	rswitch2_serdes_write32(rsw2->serdes_base_addr, VR_XS_PCS_DIG_CTRL1, BANK_380, 0x8000);
 
+	/* Re-Initialize SRAM */
+	ret = rswitch2_serdes_init_sram(rsw2);
+	if (ret)
+		return ret;
 
+	/* Check for soft reset done */
+	ret = rswitch2_serdes_reg_wait(rsw2->serdes_base_addr,
+			SR_XS_PCS_CTRL1, BANK_380, BIT(15), 0);
+	if (ret) {
+		dev_err(rsw2->dev, "SERDES does not finished soft reset (step 8)\n");
+		return ret;
+	}
 	return 0;
 }
 
@@ -408,19 +508,11 @@ static int rswitch2_serdes_chan_setting(struct rswitch2_eth_port *eth_port, enum
 
 	addr = phy_port->serdes_chan_addr;
 
-
+	//printk("%s(port=%d, mode=%d)\n", __FUNCTION__, eth_port->port_num, mode);
 	switch (mode) {
 	case SGMII:
-		rswitch2_serdes_write32(addr, SR_XS_PCS_CTRL2, BANK_300, 0x01);
 		rswitch2_serdes_write32(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, 0x2000);
-
-		/* Set common settings*/
-		ret = rswitch2_serdes_common_setting(eth_port);
-		if (ret)
-			return ret;
-
-		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_MPLL_CMN_CTRL,
-				       BANK_180, 0x11);
+		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_MPLL_CMN_CTRL, BANK_180, 0x11);
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_VCO_CAL_LD0, BANK_180, 0x540);
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_VCO_CAL_REF0, BANK_180, 0x15);
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_RX_GENCTRL1, BANK_180, 0x100);
@@ -436,22 +528,38 @@ static int rswitch2_serdes_chan_setting(struct rswitch2_eth_port *eth_port, enum
 
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_TX_GENCTRL2, BANK_180, 0x0101);
 		ret = rswitch2_serdes_reg_wait(addr, VR_XS_PMA_MP_12G_16G_TX_GENCTRL2, BANK_180, BIT(0), 0);
-		if (ret)
+		if (ret) {
+			printk("%s %d failed\n", __FUNCTION__, __LINE__);
 			return ret;
-
+		}
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_RX_GENCTRL2, BANK_180, 0x101);
 		ret = rswitch2_serdes_reg_wait(addr, VR_XS_PMA_MP_12G_16G_RX_GENCTRL2, BANK_180, BIT(0), 0);
-		if (ret)
+		if (ret) {
+			printk("%s %d failed\n", __FUNCTION__, __LINE__);
 			return ret;
+		}
 
+		//Accept AutoNeg
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_TX_GENCTRL1, BANK_180, 0x1310);
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_TX_EQ_CTRL0, BANK_180, 0x1800);
 		rswitch2_serdes_write32(addr, VR_XS_PMA_MP_12G_16G_25G_TX_EQ_CTRL1, BANK_180, 0);
 
+		rswitch2_serdes_write32(addr, SR_XS_PCS_CTRL2, BANK_300, 0x01);
+
 		rswitch2_serdes_write32(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, 0x2100);
 		ret = rswitch2_serdes_reg_wait(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, BIT(8), 0);
-		if (ret)
+		if (ret) {
+			printk("%s %d failed\n", __FUNCTION__, __LINE__);
 			return ret;
+		}
+#if 0
+		//Enable AutoNeg
+		rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x0140);  //start from 1000
+		//rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x2100);  //start from 100
+		rswitch2_serdes_write32(addr, 0x004, BANK_1F00, 5);
+		rswitch2_serdes_write32(addr, 0x028, BANK_1F00, 0x7a1);
+		rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x208);
+#endif
 		break;
 
 	case USXGMII:
@@ -527,11 +635,20 @@ static int rswitch2_serdes_set_speed(struct rswitch2_drv *rsw2, enum rswitch_ser
 	switch (mode) {
 	case SGMII:
 		if (speed == 1000)
-			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x140);
+			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x0140);
 		else if (speed == 100)
 			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x2100);
 		else if (speed == 10)
-			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x100);
+			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x0100);
+		else
+			return -EOPNOTSUPP;
+#if 0
+		//Autoneg for 1G
+		ret = rswitch2_serdes_reg_wait(addr, 0x008, BANK_1F00, BIT(0), 1);
+		if (ret)
+			return ret;
+		rswitch2_serdes_write32(addr, 0x008, BANK_1F00, 0);
+#endif
 		break;
 
 	case USXGMII:
@@ -539,11 +656,11 @@ static int rswitch2_serdes_set_speed(struct rswitch2_drv *rsw2, enum rswitch_ser
 			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x120);
 		else
 			return -EOPNOTSUPP;
-		udelay(20);
-		rswitch2_serdes_write32(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, 0x2600);
-		ret = rswitch2_serdes_reg_wait(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, BIT(10), 0);
-		if (ret)
-			return ret;
+	//	udelay(50);
+	//	rswitch2_serdes_write32(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, 0x2600);
+	//	ret = rswitch2_serdes_reg_wait(addr, VR_XS_PCS_DIG_CTRL1, BANK_380, BIT(10), 0);
+	//	if (ret)
+	//		return ret;
 		break;
 
 	default:
@@ -553,6 +670,11 @@ static int rswitch2_serdes_set_speed(struct rswitch2_drv *rsw2, enum rswitch_ser
 	return 0;
 }
 
+static int dummy_config_aneg(struct phy_device *phydev)
+{
+	printk("%s:%d phydev=%p\n", __FUNCTION__, __LINE__, phydev);
+	return 0;
+}
 
 static int rswitch2_serdes_init(struct net_device *ndev, unsigned int port_num)
 {
@@ -573,72 +695,57 @@ static int rswitch2_serdes_init(struct net_device *ndev, unsigned int port_num)
 
 	/* TODO: Support more modes and speed selection */
 	printk("%s: sw0p%d  phy_mode=%d\n", __FUNCTION__, port_num, ndev->phydev->interface );
-	switch (ndev->phydev->interface ) {
+	switch (ndev->phydev->interface) {
 	case PHY_INTERFACE_MODE_SGMII:
 		mode = SGMII;
 		speed = 1000;
-		printk("%s:%d Mode is SGMII æ %dmbit/s\n", __FUNCTION__, __LINE__, speed);
-
+		netdev_info(ndev, "port %d uses SGMII at %d Mbit/s\n", port_num, speed);
 		break;
 	case PHY_INTERFACE_MODE_USXGMII:
 		mode = USXGMII;
 		speed = 2500;
-		printk("%s:%d Mode is USXGMII æ %dmbit/s\n", __FUNCTION__, __LINE__, speed);
+		netdev_info(ndev, "port %d uses USXGMII at %d Mbit/s\n", port_num, speed);
+
+		//ugly hack to prevent auto neg by generic driver
+		if (ndev->phydev->drv->config_aneg == NULL)
+			ndev->phydev->drv->config_aneg = &dummy_config_aneg;
+
 		break;
 	default:
-		pr_debug("%s: Don't support this interface", __func__);
+		pr_debug("%s: Don't support this interface %d on port %d", __func__,
+			ndev->phydev->interface, port_num);
 		return -EOPNOTSUPP;
 	}
 
 
-
 	if (!rsw2->serdes_common_init_done) {
-
 		rsw2->serdes_common_init_done = true;
 
-		/* Disable FUSE_OVERRIDE_EN */
-		if (ioread32(phy_port->serdes_chan_addr + RSWITCH_SERDES_FUSE_OVERRIDE(port_num)))
-			iowrite32(0, phy_port->serdes_chan_addr + RSWITCH_SERDES_FUSE_OVERRIDE(port_num));
-
-		/* Initialize SRAM */
-		ret = rswitch2_serdes_init_sram(rsw2);
-		if (ret)
-			return ret;
-
-		//printk("Step 2\n");
-		rswitch2_serdes_write32(phy_port->serdes_chan_addr, VR_XS_PCS_SFTY_MR_CTRL, BANK_380, 0x443);
-
 		/* Set common settings*/
 		//printk("Step 3 (U.4)\n");
 		ret = rswitch2_serdes_common_setting(eth_port);
 		if (ret)
 			return ret;
-
-		/* Assert softreset for PHY */
-		//printk("Step 4,5\n");
-		rswitch2_serdes_write32(phy_port->serdes_chan_addr, 0x03d0, BANK_380, 1);
-		rswitch2_serdes_write32(rsw2->serdes_base_addr, VR_XS_PCS_DIG_CTRL1, BANK_380, 0x8000);
-		/* Re-Initialize SRAM */
-		//printk("Step 6,7,8\n");
-		/* Initialize SRAM */
-		ret = rswitch2_serdes_init_sram(rsw2);
-		if (ret)
-			return ret;
 	}
 
 
 	/* Set channel settings*/
 	ret = rswitch2_serdes_chan_setting(eth_port, mode);
-	if (ret)
+	if (ret) {
+		netdev_err(ndev, "channel specific SERDES configuration for port %d failed. ret=%d\n", port_num, ret);
 		return ret;
+	}
 
 	/* Set speed (bps) */
 	ret = rswitch2_serdes_set_speed(rsw2, mode, speed);
-	if (ret)
+	if (ret) {
+		netdev_err(ndev, "set initial SERDES speed for port %d failed\n", port_num);
 		return ret;
+	}
 
+#if 0
 	/* The serdes connection to PHYs takes quite long */
-	ret = rswitch2_serdes_reg_wait(rsw2->serdes_base_addr, SR_XS_PCS_STS1, BANK_300, BIT(2), BIT(2));
+	ret = rswitch2_serdes_reg_wait(phy_port->serdes_chan_addr, SR_XS_PCS_STS1, BANK_300, BIT(2), BIT(2));
 	printk("%s:%d link-up ret=%d\n", __FUNCTION__, __LINE__, ret);
 	if (ret) {
 		//reset RX side and retry
@@ -647,22 +754,24 @@ static int rswitch2_serdes_init(struct net_device *ndev, unsigned int port_num)
 		udelay(10);
 		rswitch2_serdes_write32(phy_port->serdes_chan_addr, VR_XS_PMA_MP_12G_16G_25G_RX_GENCTRL1, BANK_180, value);
 
-		ret = rswitch2_serdes_reg_wait(rsw2->serdes_base_addr, SR_XS_PCS_STS1, BANK_300, BIT(2), BIT(2));
-		//printk("%s:%d link-up ret=%d\n", __FUNCTION__, __LINE__, ret);
+		ret = rswitch2_serdes_reg_wait(phy_port->serdes_chan_addr, SR_XS_PCS_STS1, BANK_300, BIT(2), BIT(2));
+		printk("%s:%d link-up ret=%d\n", __FUNCTION__, __LINE__, ret);
 	}
 	if (ret) {
-		pr_debug("\n%s: \033[31mSerDes has not yet reached link up\033[m", __func__);
-		return ret;
+		netdev_err(ndev, "SerDes on port %d has not reached link up\n", port_num);
+		//return ret;
 	}
+	else
+		dev_info(rsw2->dev, "SerDes on port %d connected.\n", port_num);
+#endif
 
 	//printk("Step 11,12\n");
 	rswitch2_serdes_write32(phy_port->serdes_chan_addr, 0x03c0, BANK_380, 0);
 	rswitch2_serdes_write32(phy_port->serdes_chan_addr, 0x03d0, BANK_380, 0);
 
-
-
 	return 0;
 }
+
 static void rswitch2_phy_port_update_stats(struct rswitch2_physical_port *phy_port) {
 
 	u64 bytes_upper;
@@ -702,6 +811,62 @@ static void rswitch2_stat_timer(struct timer_list *t)
 
 }
 
+/*static int rswitch2_mac_set_speed(struct rswitch2_eth_port *eth_port)
+{
+	u32 oldval, newval, mask;
+	int ret;
+	struct rswitch2_physical_port *phy_port;
+
+	phy_port = eth_port->phy_port;
+
+	//TODO code the correct speeds
+
+	newval = 0;
+	switch (phy_port->phy_iface) {
+		case PHY_INTERFACE_MODE_SGMII :
+			//speed may chage during link up. MII is fixed
+			newval |= FIELD_PREP(MPIC_LSC, rsw2_rmac_1000mbps);
+			newval |= FIELD_PREP(MPIC_PIS, rsw2_rmac_gmii);
+			break;
+		case PHY_INTERFACE_MODE_USXGMII :
+			//speed may chage during link up. MII is fixed
+			newval |= FIELD_PREP(MPIC_LSC, rsw2_rmac_2500mbps);
+			newval |= FIELD_PREP(MPIC_PIS, rsw2_rmac_xgmii);
+			break;
+		default:
+			netdev_err(eth_port->ndev, "Unsupported MAC xMII format %s (%d)\n",phy_modes(phy_port->phy_iface), phy_port->phy_iface );
+			return -EINVAL;
+	}
+
+	oldval = ioread32(phy_port->rmac_base_addr + RSW2_RMAC_MPIC);
+	mask = MPIC_LSC | MPIC_PIS;
+	if ((oldval & mask) != (newval & mask)) {
+		printk("old mode is %x\n", ioread32(eth_port->ndev->base_addr + RSW2_ETHA_EAMC));
+
+		ret = rswitch2_emac_set_state(eth_port->ndev, emac_disable);
+		if (ret < 0)
+			return ret;
+		ret = rswitch2_emac_set_state(eth_port->ndev, emac_config);
+		if (ret < 0)
+			return ret;
+
+		newval |= oldval & ~mask;
+		printk("[%x] = %x\n", (u32)(phy_port->rmac_base_addr), newval);
+		iowrite32(newval, phy_port->rmac_base_addr + RSW2_RMAC_MPIC);
+
+		oldval = ioread32(phy_port->rmac_base_addr + RSW2_RMAC_MPIC);
+		netdev_info(eth_port->ndev, "Configure MAC to %x\n", oldval);
+
+		ret = rswitch2_emac_set_state(eth_port->ndev, emac_disable);
+		if (ret < 0)
+			return ret;
+		ret = rswitch2_emac_set_state(eth_port->ndev, emac_operation);
+		if (ret < 0)
+			return ret;
+	}
+	return 0;
+}*/
+
 static int rswitch2_eth_open(struct net_device *ndev)
 {
 	struct rswitch2_eth_port *eth_port;
@@ -733,6 +898,10 @@ static int rswitch2_eth_open(struct net_device *ndev)
 			return ret;
 		}
 
+		//ret = rswitch2_mac_set_speed(eth_port);
+		//if (ret != 0)
+		//	return ret;
+
 		ret = rswitch2_serdes_init(ndev, phy_port_num);
 		if (ret != 0) {
 			netdev_err(ndev, "rswitch2_serdes_init failed: %d\n", ret);
@@ -740,7 +909,6 @@ static int rswitch2_eth_open(struct net_device *ndev)
 		}
 		else {
 			netdev_err(ndev, "rswitch2_serdes_init SUCCESS\n");
-
 		}
 
 		timer_setup(&eth_port->stat_timer, rswitch2_stat_timer, 0);
@@ -1013,8 +1181,8 @@ static netdev_tx_t rswitch2_eth_start_xmit(struct sk_buff *skb, struct net_devic
 	intern_port = eth_port->intern_port;
 
 	/* Only internal ports can TX */
-	if(intern_port == NULL) {
-		netdev_dbg(ndev, "Transmission on physical port (Not implemented\n");
+	if (intern_port == NULL) {
+		//netdev_warn(ndev, "Direct transmission on physical port not yet implemented\n");
 		return NETDEV_TX_OK;
 	}
 
@@ -1676,8 +1844,10 @@ static const struct ethtool_ops rswitch2_ethtool_ops = {
 static void rswitch2_init_port_mac(struct net_device *port_ndev)
 {
 	u32 reg_val;
+	int speed;
 	struct rswitch2_eth_port *eth_port;
 	struct rswitch2_physical_port *phy_port;
+	phy_interface_t phy_iface;
 
 	eth_port = netdev_priv(port_ndev);
 	phy_port = eth_port->phy_port;
@@ -1689,22 +1859,37 @@ static void rswitch2_init_port_mac(struct net_device *port_ndev)
 	/* Set clock divider */
 	reg_val |= FIELD_PREP(MPIC_PSMCS, 0x40);
 
-	/* FIXME: Use passed values */
-	/* Set xMII type */
-	//rswitch_etha_write(etha, MPIC_PIS_GMII | MPIC_LSC_1G, MPIC);
-
+	/* speed for MAC will later updated when link comes up */
+	//phy_iface = phy_port->phy_iface;  //not yet initialised here
+	phy_iface = eth_port->rsw2->port_data[eth_port->port_num-1].phy_iface;
+	switch (phy_iface) {
+		case PHY_INTERFACE_MODE_SGMII :
+			//speed may chage during link up. MII is fixed
+			speed = 1000;
+			reg_val |= FIELD_PREP(MPIC_LSC, rsw2_rmac_1000mbps);
+			reg_val |= FIELD_PREP(MPIC_PIS, rsw2_rmac_gmii);
+			break;
+		case PHY_INTERFACE_MODE_USXGMII :
+			//speed may chage during link up. MII is fixed
+			speed = 2500;
+			reg_val |= FIELD_PREP(MPIC_LSC, rsw2_rmac_2500mbps);
+			reg_val |= FIELD_PREP(MPIC_PIS, rsw2_rmac_xgmii);
+			break;
+		default:
+			dev_err(eth_port->rsw2->dev, "Unsupported MAC xMII format %s (%d) on port %d\n",phy_modes(phy_iface), phy_iface, eth_port->port_num-1);
+			//return -EINVAL;
+	}
+	dev_info(eth_port->rsw2->dev, "etha%d uses %s at %d Mbps\n", eth_port->port_num-1, phy_modes(phy_iface), speed);
 
-	reg_val |= FIELD_PREP(MPIC_LSC, rsw2_rmac_1000mbps);
-	reg_val |= FIELD_PREP(MPIC_PIS, rsw2_rmac_gmii);
 	iowrite32(reg_val, phy_port->rmac_base_addr + RSW2_RMAC_MPIC);
 
-
+	//TODO: This is part of the MDIO access function, pre-conf has no effect
 	reg_val = ioread32(phy_port->rmac_base_addr + RSW2_RMAC_MPSM);
 	reg_val |= MPSM_MFF;
 	iowrite32(reg_val, phy_port->rmac_base_addr + RSW2_RMAC_MPSM);
 
 
-	/* Enable broad-/multi-/uni-cast */
+	/* Enable broad-/multi-/uni-cast reception of eMAC and pMAC frames*/
 	reg_val = MRAFC_BCENE;
 	reg_val |= MRAFC_MCENE;
 	reg_val |= MRAFC_UCENE;
@@ -1781,7 +1966,7 @@ static void rswitch2_init_mac_addr(struct net_device *ndev)
 	eth_port = netdev_priv(ndev);
 	rsw2 = eth_port->rsw2;
 
-	dev_info(rsw2->dev, "%s:%d: Port: %d\n", __FUNCTION__, __LINE__, eth_port->port_num);
+	//dev_info(rsw2->dev, "%s:%d: Port: %d\n", __FUNCTION__, __LINE__, eth_port->port_num);
 
 	if (!eth_platform_get_mac_address(&ndev->dev, ndev->dev_addr)) {
 		if (is_valid_ether_addr(ndev->dev_addr)) {
@@ -1792,15 +1977,13 @@ static void rswitch2_init_mac_addr(struct net_device *ndev)
 
 	} else {
 		struct rswitch2_port_data *port_data;
-
-		dev_info(rsw2->dev, "%s:%d: Using static port data\n", __FUNCTION__, __LINE__);
-
+		//dev_info(rsw2->dev, "%s:%d: Using static port data\n", __FUNCTION__, __LINE__);
 
 		port_data = &rsw2->port_data[eth_port->port_num];
 
 		/* Use MAC address from port_data */
 		memcpy(ndev->dev_addr, port_data->mac_addr, ndev->addr_len);
-		netdev_info(ndev, "Assigned MAC %.2X:%.2X:%.2X:%.2X:%.2X:%.2X to '%s'\n",
+		dev_info(rsw2->dev, "Assigned MAC %.2X:%.2X:%.2X:%.2X:%.2X:%.2X to '%s'\n",
 				port_data->mac_addr[0], port_data->mac_addr[1],
 				port_data->mac_addr[2], port_data->mac_addr[3],
 				port_data->mac_addr[4], port_data->mac_addr[5],
@@ -1822,12 +2005,12 @@ static struct device_node *rswitch2_get_port_node(struct net_device *ndev, unsig
 	eth_port = netdev_priv(ndev);
 	rsw2 = eth_port->rsw2;
 
-	dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
+	//dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 	ports = of_get_child_by_name(rsw2->dev->of_node, "ports");
 	if (!ports)
 		return NULL;
 
-	dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
+	//dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 	for_each_child_of_node(ports, port) {
 		err = of_property_read_u32(port, "reg", &index);
 		if (err < 0)
@@ -1835,9 +2018,9 @@ static struct device_node *rswitch2_get_port_node(struct net_device *ndev, unsig
 		if (index == port_num)
 			break;
 	}
-	dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
+	//dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 	of_node_put(ports);
-	dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
+	//dev_info(rsw2->dev, "dbg line: %d\n", __LINE__);
 	dev_info(rsw2->dev, "Got port node: 0x%px\n", port);
 
 
@@ -1941,6 +2124,8 @@ static int rswitch2_mdio_read(struct mii_bus *bus, int addr, int regnum)
 	struct net_device *ndev = bus->priv;
 	int mode, devad, regad;
 
+	//printk("rswitch2_mdio_read(%lx, %x, %x)\n",  (long unsigned)((struct rswitch2_eth_port *)netdev_priv(ndev))->phy_port->rmac_base_addr, addr, regnum);
+
 	mode = regnum & MII_ADDR_C45;
 	devad = (regnum >> MII_DEVADDR_C45_SHIFT) & 0x1f;
 	regad = regnum & MII_REGADDR_C45_MASK;
@@ -1948,8 +2133,8 @@ static int rswitch2_mdio_read(struct mii_bus *bus, int addr, int regnum)
 	/* Not support Clause 22 access method */
 	if (!mode) {
 		// FIXME
-		printk("%s(): NOT C45\n", __FUNCTION__);
-//		return 0;
+		printk("%s(): NOT C45 regnum=%04x  return 0\n", __FUNCTION__, regnum);
+		return 0;
 	}
 	return rswitch2_mdio_access(ndev, true, addr, devad, regad, 0);
 }
@@ -1959,6 +2144,8 @@ static int rswitch2_mdio_write(struct mii_bus *bus, int addr, int regnum, u16 va
 	struct net_device *ndev = bus->priv;
 	int mode, devad, regad;
 
+	//printk("rswitch2_mdio_write(%lx, %x, %x, %04x)\n",   (long unsigned)((struct rswitch2_eth_port *)netdev_priv(ndev))->phy_port->rmac_base_addr, addr, regnum, val);
+
 	mode = regnum & MII_ADDR_C45;
 	devad = (regnum >> MII_DEVADDR_C45_SHIFT) & 0x1f;
 	regad = regnum & MII_REGADDR_C45_MASK;
@@ -1966,8 +2153,8 @@ static int rswitch2_mdio_write(struct mii_bus *bus, int addr, int regnum, u16 va
 	/* Not support Clause 22 access method */
 	if (!mode) {
 		// FIXME
-		printk("%s(): NOT C45\n", __FUNCTION__);
-//		return 0;
+		printk("%s(): NOT C45 ignore write\n", __FUNCTION__);
+		return 0;
 	}
 
 	return rswitch2_mdio_access(ndev, false, addr, devad, regad, val);
@@ -2003,7 +2190,7 @@ static int rswitch2_mdio_init(struct net_device *ndev, unsigned int port_num)
 
 
 	mii_bus->name = "rswitch_mii";
-	sprintf(mii_bus->id, "etha%d", port_num);
+	sprintf(mii_bus->id, RSW2_PORT_BASENAME, port_num);
 	mii_bus->priv = ndev;
 	mii_bus->read = rswitch2_mdio_read;
 	mii_bus->write = rswitch2_mdio_write;
@@ -2068,8 +2255,10 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 
 
 	strncpy(port_ndev->name, RSW2_NETDEV_BASENAME, sizeof(port_ndev->name) - sizeof(port_name) - 1);
-	port_name_len = snprintf(port_name, sizeof(port_name), "p%1u", eth_port->port_num);
+	port_name_len = snprintf(port_name, sizeof(port_name), "p%1u", eth_port->port_num-1);
 	strncat(port_ndev->name, port_name, port_name_len);
+	//if (rsw2->dev->init_name == NULL)
+	//	rsw2->dev->init_name = port_ndev->name;
 
 	ret = rswitch2_emac_set_state(port_ndev, emac_disable);
 	if (ret < 0) {
@@ -2079,7 +2268,7 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 
 	ret = rswitch2_emac_set_state(port_ndev, emac_config);
 	if (ret < 0) {
-		dev_err(rsw2->dev, "Port set state 'config' failed\n");
+		netdev_err(port_ndev, "Port set state 'config' failed\n");
 		goto cleanup_phy_port;
 	}
 
@@ -2088,18 +2277,46 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 	rswitch2_port_set_mac_addr(port_ndev);
 	rswitch2_init_port_mac(port_ndev);
 
+	/*ret = rswitch2_emac_set_state(port_ndev, emac_disable);
+	if (ret < 0) {
+		dev_err(rsw2->dev, "Port set state 'disable' failed\n");
+		goto cleanup_phy_port;
+	}*/
 
-	/* Change to OPERATION Mode */
+/* Change to OPERATION Mode */
 	ret = rswitch2_emac_set_state(port_ndev, emac_operation);
 	if (ret < 0) {
-		dev_err(rsw2->dev, "Port set state failed\n");
+		netdev_err(port_ndev, "Port set state 'operation' failed\n");
 		goto cleanup_phy_port;
 	}
 
+/*
+	ret = rswitch2_emac_set_state(port_ndev, emac_disable);
+	if (ret < 0) {
+		dev_err(rsw2->dev, "Port set state 'disable' failed\n");
+		goto cleanup_phy_port;
+	}
+
+	ret = rswitch2_emac_set_state(port_ndev, emac_config);
+	if (ret < 0) {
+		netdev_err(port_ndev, "Port set state 'config' failed\n");
+		goto cleanup_phy_port;
+	}
+	ret = rswitch2_emac_set_state(port_ndev, emac_operation);
+	if (ret < 0) {
+		netdev_err(port_ndev, "Port set state 'operation' failed\n");
+		goto cleanup_phy_port;
+	}
+*/
+
 	/* Link Verification */
-	{
+	/* TODO CMARD: I'm not sure if this is the correct position. The link verification
+	 * is a feature to ask the far end MAC if it is pre-emption capable. But this can
+	 * only happen when the physical link is established.
+	 */
+	/*{
 		u32 reg_val;
-		/* Request Link Verification */
+		/ * Request Link Verification * /
 		reg_val = ioread32(phy_port->rmac_base_addr + RSW2_RMAC_MPSM);
 			reg_val |= MPSM_MFF;
 			iowrite32(reg_val, phy_port->rmac_base_addr + RSW2_RMAC_MPSM);
@@ -2110,12 +2327,12 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 							reg_val & MLVC_PLV,
 							RSWITCH2_REG_POLL_DELAY, RSWITCH2_REG_POLL_TIMEOUT);
 		if (ret != 0) {
-			dev_err(rsw2->dev, "RMAC Link verification timed out\n");
+			dev_err(rsw2->dev, "RMAC Preemption link verification timed out\n");
 			return ret;
 		} else {
-			dev_info(rsw2->dev, "RMAC Link verified!\n");
+			dev_info(rsw2->dev, "RMAC Preemption link verified!\n");
 		}
-	}
+	}*/
 
 	port_ndev->min_mtu = 2048 - (ETH_HLEN + VLAN_HLEN + ETH_FCS_LEN);
 	port_ndev->max_mtu = ETH_MIN_MTU;
@@ -2195,7 +2412,7 @@ static int rswitch2_init_internal_port(struct rswitch2_drv *rsw2, unsigned int p
 
 	intern_port = kzalloc(sizeof(struct rswitch2_internal_port), GFP_KERNEL);
 	if (intern_port < 0) {
-		dev_err(rsw2->dev, "Failed to allocate physical port\n");
+		dev_err(rsw2->dev, "Failed to allocate internal port\n");
 		ret = -ENOMEM;
 		goto intern_port_err;
 	}
@@ -2208,7 +2425,7 @@ static int rswitch2_init_internal_port(struct rswitch2_drv *rsw2, unsigned int p
 	eth_port->intern_port = intern_port;
 
 	/* Get MAC address and set it in HW */
-	dev_info(rsw2->dev, "Before init MAC\n");
+	dev_info(rsw2->dev, "Start init the MAC\n");
 
 	rswitch2_init_mac_addr(ndev);
 	rswitch2_intern_set_mac_addr(ndev);
@@ -2421,6 +2638,9 @@ int rswitch2_eth_init(struct rswitch2_drv *rsw2)
 		}
 	}
 
+	/* the common SERDES initialisation for RSW2/PHY connection */
+	//rswitch2_serdes_common_init(rsw2);
+
 	for (cur_port = 0; cur_port < rsw2->num_of_tsn_ports; cur_port++) {
 		rswitch2_init_physical_port(rsw2, cur_port);
 		if (ret < 0) {
diff --git a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.h b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.h
index 8514ae368511..325059a1248a 100644
--- a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.h
+++ b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.h
@@ -14,6 +14,7 @@
 
 
 #define RSW2_NETDEV_BASENAME	"sw%d"
+#define RSW2_PORT_BASENAME	    "etha%d"
 
 #define BE_TX_RING_MAX	1024
 #define BE_RX_RING_MAX	2048
diff --git a/drivers/net/ethernet/renesas/rswitch2/rswitch2_platf.c b/drivers/net/ethernet/renesas/rswitch2/rswitch2_platf.c
index 7e52a48929cd..37fd9a0c1fef 100644
--- a/drivers/net/ethernet/renesas/rswitch2/rswitch2_platf.c
+++ b/drivers/net/ethernet/renesas/rswitch2/rswitch2_platf.c
@@ -161,7 +161,7 @@ static phy_interface_t rswitch2_get_phy_inferface(const char *phy_mode_str) {
 	for(i = 0; i < elem; i++) {
 		int ret;
 		ret = strcmp(phy_mode_str, rsw2_phy_mode_xlate_tbl[i].dt_str);
-		printk("Comparing '%s' <-> '%s'\n", phy_mode_str, rsw2_phy_mode_xlate_tbl[i].dt_str);
+		//printk("Comparing '%s' <-> '%s'\n", phy_mode_str, rsw2_phy_mode_xlate_tbl[i].dt_str);
 		if(ret == 0) {
 			intf = rsw2_phy_mode_xlate_tbl[i].intf;
 			break;
@@ -240,11 +240,11 @@ static int rswitch2_platf_set_port_data(struct rswitch2_drv *rsw2)
 	ports = of_get_child_by_name(rsw2->dev->of_node, "ports");
 	if (!ports) {
 		dev_err(rsw2->dev, "No ports specified in device tree\n");
-//		return -EINVAL;
-	}
-	else {
-		dev_err(rsw2->dev, "Got ports from node\n");
+		return -EINVAL;
 	}
+	//else {
+	//	dev_err(rsw2->dev, "Got ports from device tree nodes\n");
+	//}
 
 
 	dt_port_num = of_get_child_count(ports);
@@ -254,7 +254,7 @@ static int rswitch2_platf_set_port_data(struct rswitch2_drv *rsw2)
 		return -EINVAL;
 	}
 	else {
-		dev_info(rsw2->dev, "DT specified %d ports\n", dt_port_num);
+		dev_info(rsw2->dev, "DT specifies %d Ethernet ports\n", dt_port_num);
 	}
 
 
@@ -267,25 +267,25 @@ static int rswitch2_platf_set_port_data(struct rswitch2_drv *rsw2)
 		//cur_port_data->phy_id;
 
 		ret = of_property_read_u32(port, "reg", &cur_port_data->reg);
-		if(ret < 0) {
+		if (ret < 0) {
 			dev_err(rsw2->dev, "Failed to get register property for port %d: %d\n",
 					cur_port_num, ret);
 			break;
 		}
-		else {
-			dev_info(rsw2->dev, "Got reg=%u for port %d\n", cur_port_num, cur_port_data->reg);
-		}
+		//else {
+		//	dev_info(rsw2->dev, "Got reg=%u for port %d\n", cur_port_data->reg, cur_port_num);
+		//}
 
 		ret = of_property_read_string(port, "phy-mode", &prop_phy_mode_str);
-		if(ret < 0) {
+		if (ret < 0) {
 			dev_err(rsw2->dev, "Failed to get phy mode property for port %d: %d\n",
 					cur_port_num, ret);
 			break;
 		}
-		dev_err(rsw2->dev, "DT phy mode: '%s'\n", prop_phy_mode_str);
+		dev_err(rsw2->dev, "DT phy mode: etha%u '%s'\n", cur_port_data->reg, prop_phy_mode_str);
 
 		cur_port_data->phy_iface = rswitch2_get_phy_inferface(prop_phy_mode_str);
-		if(cur_port_data->phy_iface == PHY_INTERFACE_MODE_NA) {
+		if (cur_port_data->phy_iface == PHY_INTERFACE_MODE_NA) {
 			dev_err(rsw2->dev, "Invalid phy mode provided for port %d\n", cur_port_num);
 			ret = -EINVAL;
 		}
-- 
2.17.1

